Sources/MlemMiddleware/FeedLoaders/Generics/MultiFetcher.swift

```diff
@@ -0,0 +1,88 @@
+//
+//  MultiFetcher.swift
+//  MlemMiddleware
+//
+//  Created by Eric Andrews on 2024-11-24.
+//
+
+class MultiFetcher<Item: FeedLoadable>: Fetcher<Item> {
+    var sources: [ChildFeedLoader<Item>]
+    var sortType: FeedLoaderSort.SortType
+    
+    init(api: ApiClient, pageSize: Int, sources: [ChildFeedLoader<Item>], sortType: FeedLoaderSort.SortType) {
+        self.sources = sources
+        self.sortType = sortType
+        
+        super.init(api: api, pageSize: pageSize)
+    }
+    
+    override func fetch() async throws -> LoadingResponse<Item> {
+        var newItems: [Item] = .init()
+        
+        while newItems.count < pageSize {
+            if let nextItem = try await computeNextItem() {
+                newItems.append(nextItem)
+            } else {
+                print("[\(Item.self) MultiFetcher] no next item found")
+                return .done(newItems)
+            }
+        }
+        
+        return .success(newItems)
+    }
+    
+    override func reset() async {
+        for source in sources {
+            await source.clear(clearParent: false)
+            print("[\(Item.self) MultiFetcher] source cleared (\(source.loadingState))")
+        }
+        
+        await super.reset()
+    }
+    
+    /// Computes and returns the highest sorted item from the tops of all sources
+    private func computeNextItem() async throws -> Item? {
+        var sortVal: FeedLoaderSort?
+        var sourceToConsume: ChildFeedLoader<Item>?
+        
+        // find the highest-sorted item from the tops of all sources
+        for source in sources {
+            (sortVal, sourceToConsume) = try await compareNextItem(lhsVal: sortVal, lhsSource: sourceToConsume, rhsSource: source)
+        }
+        
+        if let nextItem = sourceToConsume?.consumeNextItem() {
+            // ensure the item is of the correct type
+            assert(nextItem is Item, "Could not convert item to [\(Item.self)]")
+            return nextItem as! Item // this only fails if there's a serious type issue, which should be caught by the assertion above
```

Comment:
Might it be better to use `as?` here to avoid a fatal crash? The function can return `nil` anyways

```suggestion
            return nextItem as? Item
```

Reply:
Nice catch! Turns out the whole cast is unnecessary now that child loaders just use the parent type 😁